File: problem437.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def pathSum(self, root: Optional[TreeNode], targetSum: int) -> int:
        complements = defaultdict(int)
        complements[0]=1
        counts = 0
        def dfs(node, currsum):
            nonlocal complements, targetSum, counts
            if not node:
                return
            currsum+=node.val
            counts+=complements[currsum-targetSum]
            complements[currsum]+=1
            dfs(node.left, currsum)
            dfs(node.right, currsum)
            complements[currsum]-=1
            return
        
        dfs(root, 0)
        return counts
        
#Time: O(V+E)
#Space: O(V+E)

File: problem437.cs
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int val=0, TreeNode left=null, TreeNode right=null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
public class Solution {
    int pathWays=0;
    int targetSum;
    public int PathSum(TreeNode root, int targetSum) {
        this.targetSum=targetSum;
        Traverse(root);
        return pathWays;
    }
    
    public void Traverse(TreeNode node)
    {
        if(node==null)
            return;
        //Console.WriteLine("Dfs at root:" + node.val);
        Dfs(node, 0);
        Traverse(node.left);
        Traverse(node.right);
        
    }
    
    public void Dfs(TreeNode node, long currSum)
    {
        if(node==null)
            return;
        
        currSum+=node.val;
        
        if(currSum==targetSum)
            pathWays++;
        
        //Console.WriteLine("Visiting node:" + node.val + ", currsum:" + currSum + ",Pathways:" + pathWays);
        
        Dfs(node.left, currSum);
        Dfs(node.right, currSum);      
    }
}

//time: O(NlogN)
//space: O(logN)